aisutd/membership-portal
# feat: Add AWS Cognito & NextAuth Support

### Summary

Added support for NextAuth, AWS Cognito & other tooling to setup authentication for the AIS Membership Portal

### Details
 - Setup Cognito User Pool
 - Setup Cognito resource for permissions management --> `arn:aws:iam::<removed>:role/service-role/<removed>-SMS-Role`
 - Setup GCP Project `Membership-Portal`
 - Setup OAuth Consent Screen for `email`, `profile` and `openid` scopes
 - Setup OAuth 2.0 Credentials for External Authentication (Allows sign in with google)
 - Connect Cognito to GCP Credentials
 - Setup Authorization domain for AIS @ `https://aisutd.auth.us-east-2.amazoncognito.com`
 - Setup NextAuth & Add Cognito as a provider
 - Add Protected Page as example
 - Add Protected API endpoint as example
 - Update landing page to contain helpful information to test & debug the application

### Testing

 - Create a `.env.local` file & ask @harshasrikara  for environment variables

```
NEXTAUTH_URL=http://localhost:3000
COGNITO_CLIENT_ID=*
COGNITO_CLIENT_SECRET=*
COGNITO_DOMAIN=*
```

 - Run the following commands

```
npm install 

npm run dev
```

 - Open up `http://localhost:3000/`
 - Test being able to sign in, sign out & ensuring that the ID & email are correctly displayed
 - Test protected page `/secure` & make sure that you see `Access Denied` when trying to open it before having signed in
 - Test protected api `/api/protected-api-example` (GET / POST / anything works) and make sure that calling it from an external client like Postman/Insomnia gives a 401 error code. 

### Additional Context
The work completed as part of this PR completes the backend requirements for the AIS Membership Portal. Currently only sign in is allowed on the portal. Accounts need to be pre-registered either via the Cognito Admin API or on the AWS Console to determine which users can get access. The changes made to the frontend only serve as an example and are primarily for testing out the features. 

Cognito setup involved many one-time choices on the techniques / tooling for authenticating users. Reach out to @harshasrikara for more details. 

Closes #1
